docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+
+```wing
+bring cloud;
+
+let bucket = new cloud.Bucket();
+
+client.lift(lift { bucket: [put] });
+```
+
+In order to access the bucket, the client needs to know it should resolve the relevent information (e.g. bucket name or handle).
+This information is stored in client.context, so we need to way to both store and resolve this information.
+For example, we can store/retrieve the context from a file:
+
+```wing
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return fs.readJson("/var/run/wing-client-context.json");
+  },
+);
+
+// The consumer of the client will need to retrieve/link this data after deployment
+let clientContext = new cloud.Bucket() as "ctx";
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+If the runtime cost is acceptable, the context can even be directly resolved directly from the bucket during the client's runtime:
+
+```wing
+bring cloud;
+bring clients;
+
+let clientContext = new cloud.Bucket() as "ctx";
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return Json.parse(clientContext.get("ctx"));
+  },
+);
+
+// hydrates the context during deployment
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+In cases where reading from the filesystem is difficult and the overhead or a runtime lookup is too high, the context can be passed in via an environment variable:
+
+```wing
+bring cloud;
+bring clients;
+bring util;
+
+let bucket = new cloud.Bucket() as "shared";
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    let ctxRaw = util.getEnv("WING_CLIENT_CONTEXT");
+    let ctxDecoded = util.fromBase64(ctxRaw);
+    return Json.parse(ctxDecoded);
+  },
+);
+
+client.add(lift { bucket: [put] });
+```
+
+Regardless of the approach used to resolve the context, using the generated client will work the same way:
+
+```ts
+import { Client } from "@acme/infra-client";
+
+// performs the context resolution
+const client = new Client();
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions
```

Comment:
I'd love to tackle permissions here, but I struggled to think of a solution that doesn't explode the scope of this RFC